compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/CoroutineCodegen.kt

```diff
@@ -66,7 +67,7 @@ internal fun generateStateMachineForNamedFunction(
         shouldPreserveClassInitialization = state.constructorCallNormalizationMode.shouldPreserveClassInitialization,
         containingClassInternalName = classCodegen.visitor.thisName,
         isForNamedFunction = true,
-        needDispatchReceiver = irFunction.dispatchReceiverParameter != null,
+        needDispatchReceiver = irFunction.dispatchReceiverParameter != null || irFunction.name.asString().endsWith(SUSPEND_IMPL_NAME_SUFFIX),
```

Comment:
A non-name based way of determining this would be nice. I'm open to suggestions! :)

Reply:
You can use different origin, something like `SUSPEND_IMPL_STATIC_FUNCTION`.

Reply:
Done, I added it to IrDeclarationOrigin.kt. However, it looks like it is JVM_IR only, so it might be better to move it somewhere else?

Reply:
There's `JvmLoweredDeclarationOrigin` for JVM-specific origins

Reply:
Thanks @udalov, that is what I was looking for. I found JvmDeclarationOrigin which wasn't exactly what I was looking for. Done!